# Extraccion

Como veremos, el objetivo de este jupyter es leer las bases de datos de una manera acorde para poder guardar toda la informacion en formato Parquet.

In [1]:
import pandas as pd #pandas para trabajar sobre DataFrames
import Modulo as md
import numpy as np
np.set_printoptions(suppress=True)
pd.set_option('display.float_format', lambda x: '%.2f' % x)

ModuleNotFoundError: No module named 'Modulo'

## Funcion de descompresion y lectura

Haremos una funcion para descomprimir y leer correctamente los archivos, notando que dos de ellos tienen una estructura particular de json (comillas simples). Para ello, vamos a tener que leer el json como un diccionario de python y luego pasarlo a un dataframe.

Esta funcion se encuentra dentro del archivo Modulo.py, asi como todas las funciones soporte que usaremos a lo largo del ETL.

## Nombres de archivos

Ahora creemos una funcion para obtener los nombres de los archivos iniciales y el parquet sobre el que vamos a trabajar a posteriori (es un formato mas eficiente y practico)

## Descompresion

In [4]:
ruta = 'Datasets/' #ruta de carpeta
nombres = md.get_names(ruta) #llamamos a la primera funcion
dic = {}

for k,v in nombres.items(): #enviamos los 3 dataframes a un diccionario para poder acceder a ellos facilmente
    dic.update({k: md.gz_a_DataFrame(v[0])})

In [5]:
dic['steam_games']

publisher                                 genres  \
0                   NaN                                    NaN   
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title  \
0                            NaN                       NaN   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date  \
0                                                     NaN          NaN   
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price early_access  \
0                                        

A simple vista vemos que vamos a tener que trabajar sobre los registros nulos, lo dejaremos para despues.

In [6]:
dic['user_reviews']

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

Aca tenemos algo para trabajar, hay una columna anidada.

In [7]:
dic['users_items']

user_id  items_count           steam_id  \
0            76561197970982479          277  76561197970982479   
1                      js41637          888  76561198035864385   
2                    evcentric          137  76561198007712555   
3                   Riot-Punch          328  76561197963445855   
4                        doctr          541  76561198002099482   
...                        ...          ...                ...   
88305        76561198323066619           22  76561198323066619   
88306        76561198326700687          177  76561198326700687   
88307  XxLaughingJackClown77xX            0  76561198328759259   
88308        76561198329548331            7  76561198329548331   
88309        edward_tremethick            0  76561198331598578   

                                                user_url  \
0      http://steamcommunity.com/profiles/76561197970...   
1                   http://steamcommunity.com/id/js41637   
2                 http://steamcommunity.com/id/evcentric   
3                http://steamcommunity.com/id/Riot-Punch   
4                     http://steamcommunity.com/id/doctr   
...                                                  ...   
88305  http://steamcommunity.com/profiles/76561198323...   
88306  http://steamcommunity.com/profiles/76561198326...   
88307  http://steamcommunity.com/id/XxLaughingJackClo...   
88308  http://steamcommunity.com/profiles/76561198329...   
88309     http://steamcommunity.com/id/edward_tremethick   

                                                   items  
0      [{'item_id': '10', 'item_name': 'Counter-Strik...  
1      [{'item_id': '10', 'item_name': 'Counter-Strik...  
2      [{'item_id': '1200', 'item_name': 'Red Orchest...  
3      [{'item_id': '10', 'item_name': 'Counter-Strik...  
4      [{'item_id': '300', 'item_name': 'Day of Defea...  
...                                                  ...  
88305  [{'item_id': '413850', 'item_name': 'CS:GO Pla...  
88306  [{'item_id': '11020', 'item_name': 'TrackMania...  
88307                                                 []  
88308  [{'item_id': '304930', 'item_name': 'Unturned'...  
88309                                                 []  

[88310 rows x 5 columns]

Tenemos la misma situacion que la anterior, columna anidada.

In [8]:
columnas = {'user_reviews':'reviews','users_items':'items'} #las columnas a desanidar

for k,v in columnas.items():
    print(dic[k][v][0])

[{'funny': '', 'posted': 'Posted November 5, 2011.', 'last_edited': '', 'item_id': '1250', 'helpful': 'No ratings yet', 'recommend': True, 'review': 'Simple yet with great replayability. In my opinion does "zombie" hordes and team work better than left 4 dead plus has a global leveling system. Alot of down to earth "zombie" splattering fun for the whole family. Amazed this sort of FPS is so rare.'}, {'funny': '', 'posted': 'Posted July 15, 2011.', 'last_edited': '', 'item_id': '22200', 'helpful': 'No ratings yet', 'recommend': True, 'review': "It's unique and worth a playthrough."}, {'funny': '', 'posted': 'Posted April 21, 2011.', 'last_edited': '', 'item_id': '43110', 'helpful': 'No ratings yet', 'recommend': True, 'review': 'Great atmosphere. The gunplay can be a bit chunky at times but at the end of the day this game is definitely worth it and I hope they do a sequel...so buy the game so I get a sequel!'}]
[{'item_id': '10', 'item_name': 'Counter-Strike', 'playtime_forever': 6, 'pl

## Desanidacion de columnas

In [9]:
# verifiquemos cuantas filas deberian haber en items, aprovechando que tiene la cantidad de registros por fila anidada:

registros = dic['users_items']['items_count'].sum()

In [10]:
columnas = {'user_reviews':'reviews','users_items':'items'} #las columnas a desanidar

for k,v in columnas.items(): #iteramos sobre el diccionario
    dic[k] = dic[k].explode(v) #expandimos la lista de la columna anidada para poder leer el json
    dic[k].dropna(subset=v,inplace=True) #eliminamos los registros que no tienen informacion
    dic[k] = pd.json_normalize(dic[k].to_dict(orient='records'), sep='') #normalizamos la columna, desanidandola
    dic[k] = dic[k].rename(columns=lambda x: x.replace(v, '')) #eliminamos el prefijo del nombre de las columnas


In [11]:
filas = dic['users_items'].shape[0] #contamos la cantidad de registros para comparar

print(filas,registros,filas == registros) #comparamos los registros indicados en las columnas vs los obtenidos, no perdimos informacion

5153209 5153209 True


# Transformacion

Ahora veamos los dataframes y hagamos las transformaciones necesarias para dejar limpias las bases de trabajo.

## Steam_games

Trabajaremos sobre una variable df para luego implementar los cambios, de este modo evitamos tener que correr todo el script en caso de error u omision.

In [12]:
df = dic['steam_games']
df

publisher                                 genres  \
0                   NaN                                    NaN   
1                   NaN                                    NaN   
2                   NaN                                    NaN   
3                   NaN                                    NaN   
4                   NaN                                    NaN   
...                 ...                                    ...   
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444              NaN                                    NaN   

                        app_name                     title  \
0                            NaN                       NaN   
1                            NaN                       NaN   
2                            NaN                       NaN   
3                            NaN                       NaN   
4                            NaN                       NaN   
...                          ...                       ...   
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                       NaN   

                                                      url release_date  \
0                                                     NaN          NaN   
1                                                     NaN          NaN   
2                                                     NaN          NaN   
3                                                     NaN          NaN   
4                                                     NaN          NaN   
...                                                   ...          ...   
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                     tags  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
0                                                     NaN   
1                                                     NaN   
2                                                     NaN   
3                                                     NaN   
4                                                     NaN   
...                                                   ...   
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price early_access  \
0                                        

### Registros nulos

In [13]:
# eliminemos los registros nulos

df = df.dropna(how='all').reset_index(drop=True)
df.shape

(32135, 13)

### Registros duplicados

In [14]:
md.indentificar_duplicados(df,'id')

publisher               genres  \
74                                                   NaN                  NaN   
13894                                 Bethesda Softworks             [Action]   
14573                                 Bethesda Softworks             [Action]   
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
74                                                NaN   
13894                Wolfenstein II: The New Colossus   
14573                Wolfenstein II: The New Colossus   
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
74                                                NaN   
13894                Wolfenstein II: The New Colossus   
14573                Wolfenstein II: The New Colossus   
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date  \
74                        http://store.steampowered.com/          NaN   
13894          http://store.steampowered.com/app/612880/   2017-10-26   
14573  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26   
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags  \
74                                                   NaN   
13894  [Action, FPS, Gore, Violent, Alternate History...   
14573  [Action, FPS, Gore, Violent, Alternate History...   
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  \
74                                                   NaN   
13894  http://steamcommunity.com/app/612880/reviews/?...   
14573  http://steamcommunity.com/app/612880/reviews/?...   
30961                                                NaN   

                                                   specs price early_access  \
74                                                   NaN 19.99        False   
13894  [Single-player, Steam Achievements, Full contr... 59.99        False   
14573  [Single-player, Steam Achievements, Full contr... 59.99        False   
30961  [Single-player, Steam Achievements, Steam Trad... 19.99        False   

           id                                   developer  
74        NaN                                         NaN  
13894  612880                               Machine Games  
14573  612880                               Machine Games  
30961     NaN  Rocksteady Studios,Feral Interactive (Mac)

In [15]:
df[df['app_name'] == 'Batman: Arkham City - Game of the Year Edition']

publisher               genres  \
1068   Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]   

                                             app_name  \
1068   Batman: Arkham City - Game of the Year Edition   
30961  Batman: Arkham City - Game of the Year Edition   

                                                title  \
1068   Batman: Arkham City - Game of the Year Edition   
30961  Batman: Arkham City - Game of the Year Edition   

                                                     url release_date  \
1068   http://store.steampowered.com/app/200260/Batma...   2012-09-07   
30961           http://store.steampowered.com/app/200260   2012-09-07   

                                                    tags  \
1068   [Action, Open World, Batman, Adventure, Stealt...   
30961  [Action, Open World, Batman, Adventure, Stealt...   

                                             reviews_url  \
1068   http://steamcommunity.com/app/200260/reviews/?...   
30961                                                NaN   

                                                   specs price early_access  \
1068   [Single-player, Steam Achievements, Steam Trad... 19.99        False   
30961  [Single-player, Steam Achievements, Steam Trad... 19.99        False   

           id                                   developer  
1068   200260  Rocksteady Studios,Feral Interactive (Mac)  
30961     NaN  Rocksteady Studios,Feral Interactive (Mac)

Ya tenemos este registro con el id 200260, el mismo se encuentra igualmente completo en ambos casos asi que podemos eliminar este caso.

Tambien, el otro id Nan no nos suma informacion valiosa ya que no tienen ninguna identificacion, lo eliminaremos tambien.

In [16]:
df.dropna(subset=['id'],inplace=True)

Con respecto al otro duplicado, eliminaremos un registro ya que se encuentra duplicada la linea entera.

In [17]:
df.drop(13894,inplace=True)

### Seleccion de columnas

In [18]:
columns = ['genres','release_date','price','id','developer','app_name']

df = df[columns]

### Tipo de datos

In [19]:
md.tipo_datos(df)

Columna     Tipo_datos  %_nulos  Nulos
0        genres  [list, float]    11.38   3282
1  release_date   [str, float]     6.87   2066
2         price   [float, str]     4.48   1377
3            id          [str]     0.00      0
4     developer   [str, float]    11.44   3298
5      app_name   [str, float]     0.00      1

#### Columnas Publisher, app_name, developer

Completaremos los nulos con 'Sin_Dato'

In [20]:
columnas = ['app_name','developer']

df[columnas] = df[columnas].fillna('Sin_Dato')

md.tipo_datos(df)

Columna     Tipo_datos  %_nulos  Nulos
0        genres  [list, float]    11.38   3282
1  release_date   [str, float]     6.87   2066
2         price   [float, str]     4.48   1377
3            id          [str]     0.00      0
4     developer          [str]     0.00      0
5      app_name          [str]     0.00      0

#### Columna Price

Corrijiremos los valores str de la columna price.

In [21]:
df.price.unique()

array([4.99, 'Free To Play', 'Free to Play', 0.99, 2.99, 3.99, 9.99,
       18.99, 29.99, nan, 'Free', 10.99, 1.59, 14.99, 1.99, 59.99, 8.99,
       6.99, 7.99, 39.99, 19.99, 7.49, 12.99, 5.99, 2.49, 15.99, 1.25,
       24.99, 17.99, 61.99, 3.49, 11.99, 13.99, 'Free Demo',
       'Play for Free!', 34.99, 74.76, 1.49, 32.99, 99.99, 14.95, 69.99,
       16.99, 79.99, 49.99, 5.0, 44.99, 13.98, 29.96, 119.99, 109.99,
       149.99, 771.71, 'Install Now', 21.99, 89.99,
       'Play WARMACHINE: Tactics Demo', 0.98, 139.92, 4.29, 64.99,
       'Free Mod', 54.99, 74.99, 'Install Theme', 0.89, 'Third-party',
       0.5, 'Play Now', 299.99, 1.29, 3.0, 15.0, 5.49, 23.99, 49.0, 20.99,
       10.93, 1.39, 'Free HITMAN™ Holiday Pack', 36.99, 4.49, 2.0, 4.0,
       9.0, 234.99, 1.95, 1.5, 199.0, 189.0, 6.66, 27.99, 10.49, 129.99,
       179.0, 26.99, 399.99, 31.99, 399.0, 20.0, 40.0, 3.33, 199.99,
       22.99, 320.0, 38.85, 71.7, 59.95, 995.0, 27.49, 3.39, 6.0, 19.95,
       499.99, 16.06, 4.68, 131

In [22]:
# enviamos todos los textos a float valor 0.0 si es que no aclaran nada en el texto de valores
df['price'] = df['price'].apply(md.col_price)
df.price.unique()

array([  4.99,   0.  ,   0.99,   2.99,   3.99,   9.99,  18.99,  29.99,
        10.99,   1.59,  14.99,   1.99,  59.99,   8.99,   6.99,   7.99,
        39.99,  19.99,   7.49,  12.99,   5.99,   2.49,  15.99,   1.25,
        24.99,  17.99,  61.99,   3.49,  11.99,  13.99,  34.99,  74.76,
         1.49,  32.99,  99.99,  14.95,  69.99,  16.99,  79.99,  49.99,
         5.  ,  44.99,  13.98,  29.96, 119.99, 109.99, 149.99, 771.71,
        21.99,  89.99,   0.98, 139.92,   4.29,  64.99,  54.99,  74.99,
         0.89,   0.5 , 299.99,   1.29,   3.  ,  15.  ,   5.49,  23.99,
        49.  ,  20.99,  10.93,   1.39,  36.99,   4.49,   2.  ,   4.  ,
         9.  , 234.99,   1.95,   1.5 , 199.  , 189.  ,   6.66,  27.99,
        10.49, 129.99, 179.  ,  26.99, 399.99,  31.99, 399.  ,  20.  ,
        40.  ,   3.33, 199.99,  22.99, 320.  ,  38.85,  71.7 ,  59.95,
       995.  ,  27.49,   3.39,   6.  ,  19.95, 499.99,  16.06,   4.68,
       131.4 ,  44.98, 202.76,   1.  ,   2.3 ,   0.95, 172.24, 249.99,
      

#### Columna release_date

Corrigiremos los valores str y solo tomaremos la informacion del anio de lanzamiento, o 'Sin_Dato' si es que no ha sido lanzado.

In [23]:
df['release_year'] = df['release_date'].apply(md.release)
df['release_year'].unique()

array([2018, 2017,    0, 1997, 1998, 2016, 2006, 2005, 2003, 2007, 2002,
       2000, 1995, 1996, 1994, 2001, 1993, 2004, 1999, 2008, 2009, 1992,
       1989, 2010, 2011, 2013, 2012, 2014, 1983, 1984, 2015, 1990, 1988,
       1991, 1985, 1982, 1987, 1981, 1986, 2021, 2019, 1975, 1970, 1980],
      dtype=int64)

In [24]:
df.drop(columns='release_date',inplace=True) # drop a la columna que ya no nos sirve mas

In [25]:
df

genres  price      id  \
0          [Action, Casual, Indie, Simulation, Strategy]   4.99  761140   
1                   [Free to Play, Indie, RPG, Strategy]   0.00  643980   
2      [Casual, Free to Play, Indie, Simulation, Sports]   0.00  670290   
3                            [Action, Adventure, Casual]   0.99  767400   
4                                                    NaN   2.99  773570   
...                                                  ...    ...     ...   
32130              [Casual, Indie, Simulation, Strategy]   1.99  773640   
32131                          [Casual, Indie, Strategy]   4.99  733530   
32132                        [Indie, Racing, Simulation]   1.99  610660   
32133                                    [Casual, Indie]   4.99  658870   
32134                                                NaN   4.99  681550   

                      developer                  app_name  release_year  
0                     Kotoshiro       Lost Summoner Kitty          2018  
1              Secret Level SRL                 Ironbound          2018  
2                  Poolians.com   Real Pool 3D - Poolians          2017  
3                          彼岸领域                   弹炸人2222          2017  
4                      Sin_Dato             Log Challenge             0  
...                         ...                       ...           ...  
32130        Nikita "Ghost_RUS"            Colony On Mars          2018  
32131                    Sacada  LOGistICAL: South Africa          2018  
32132  Laush Dmitriy Sergeevich             Russian Roads          2018  
32133             xropi,stev3ns       EXIT 2 - Directions          2017  
32134                  Sin_Dato               Maze Run VR             0  

[32132 rows x 6 columns]

#### Columna Genres

Apliaremos la columna para tener una base en donde exista una linea por cada genero, repitiendose el juego.

In [26]:
df = df.explode('genres')
df.dropna(subset=['genres'],inplace=True)
df

genres  price      id                 developer  \
0          Action   4.99  761140                 Kotoshiro   
0          Casual   4.99  761140                 Kotoshiro   
0           Indie   4.99  761140                 Kotoshiro   
0      Simulation   4.99  761140                 Kotoshiro   
0        Strategy   4.99  761140                 Kotoshiro   
...           ...    ...     ...                       ...   
32132       Indie   1.99  610660  Laush Dmitriy Sergeevich   
32132      Racing   1.99  610660  Laush Dmitriy Sergeevich   
32132  Simulation   1.99  610660  Laush Dmitriy Sergeevich   
32133      Casual   4.99  658870             xropi,stev3ns   
32133       Indie   4.99  658870             xropi,stev3ns   

                  app_name  release_year  
0      Lost Summoner Kitty          2018  
0      Lost Summoner Kitty          2018  
0      Lost Summoner Kitty          2018  
0      Lost Summoner Kitty          2018  
0      Lost Summoner Kitty          2018  
...                    ...           ...  
32132        Russian Roads          2018  
32132        Russian Roads          2018  
32132        Russian Roads          2018  
32133  EXIT 2 - Directions          2017  
32133  EXIT 2 - Directions          2017  

[71551 rows x 6 columns]

#### Estandarizacion de tipeo

In [27]:
text_cols = ['genres','developer','app_name']

df[text_cols] = df[text_cols].apply(md.convertir_proper)

In [28]:
dic['steam_games'] = df.copy()

## User_reviews

In [29]:
df = dic['user_reviews']
df

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59300  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59301  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59302        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59303        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59304        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                  funny                     posted  \
0                                         Posted November 5, 2011.   
1                                            Posted July 15, 2011.   
2                                           Posted April 21, 2011.   
3                                            Posted June 24, 2014.   
4                                        Posted September 8, 2013.   
...                                 ...                        ...   
59300                                              Posted July 10.   
59301                                               Posted July 8.   
59302  1 person found this review funny             Posted July 3.   
59303                                              Posted July 20.   
59304                                               Posted July 2.   

      last_edited item_id                                          helpful  \
0                    1250                                   No ratings yet   
1                   22200                                   No ratings yet   
2                   43110                                   No ratings yet   
3                  251610  15 of 20 people (75%) found this review helpful   
4                  227300     0 of 1 people (0%) found this review helpful   
...           ...     ...                                              ...   
59300                  70                                   No ratings yet   
59301              362890                                   No ratings yet   
59302              273110    1 of 2 people (50%) found this review helpful   
59303                 730                                   No ratings yet   
59304                 440                                   No ratings yet   

       recommend                                             review  
0           True  Simple yet with great replayability. In my opi...  
1           True               It's unique and worth a playthrough.  
2           True  Great atmosphere. The gunplay can be a bit chu...  
3           True  I know what you think when you see this title ...  
4           True  For a simple (it's actually not all that simpl...  
...          ...                                                ...  
59300       True  a must have classic from steam definitely wort...  
59301       True  this game is a perfect remake of the original ...  
59302       True  had so much fun plaing this and collecting res...  
59303       True                                                 :D  
59304       True                                     so much fun :D  

[59305 rows x 9 columns]

### Registros nulos

In [30]:
# eliminemos los registros nulos
print(df.shape)
df = df.dropna(how='all').reset_index(drop=True)
print(df.shape) # no hay registros nulos completamente

(59305, 9)
(59305, 9)


### Registros duplicados

In [31]:
md.indentificar_duplicados(df,'user_id') 
# tenemos un problema, al haber ampliado las columnas los user_id estan duplicados por necesidad

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
2      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
3                js41637               http://steamcommunity.com/id/js41637   
4                js41637               http://steamcommunity.com/id/js41637   
...                  ...                                                ...   
59300  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59301  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
59302        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59303        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   
59304        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                  funny                     posted  \
0                                         Posted November 5, 2011.   
1                                            Posted July 15, 2011.   
2                                           Posted April 21, 2011.   
3                                            Posted June 24, 2014.   
4                                        Posted September 8, 2013.   
...                                 ...                        ...   
59300                                              Posted July 10.   
59301                                               Posted July 8.   
59302  1 person found this review funny             Posted July 3.   
59303                                              Posted July 20.   
59304                                               Posted July 2.   

      last_edited item_id                                          helpful  \
0                    1250                                   No ratings yet   
1                   22200                                   No ratings yet   
2                   43110                                   No ratings yet   
3                  251610  15 of 20 people (75%) found this review helpful   
4                  227300     0 of 1 people (0%) found this review helpful   
...           ...     ...                                              ...   
59300                  70                                   No ratings yet   
59301              362890                                   No ratings yet   
59302              273110    1 of 2 people (50%) found this review helpful   
59303                 730                                   No ratings yet   
59304                 440                                   No ratings yet   

       recommend                                             review  
0           True  Simple yet with great replayability. In my opi...  
1           True               It's unique and worth a playthrough.  
2           True  Great atmosphere. The gunplay can be a bit chu...  
3           True  I know what you think when you see this title ...  
4           True  For a simple (it's actually not all that simpl...  
...          ...                                                ...  
59300       True  a must have classic from steam definitely wort...  
59301       True  this game is a perfect remake of the original ...  
59302       True  had so much fun plaing this and collecting res...  
59303       True                                                 :D  
59304       True                                     so much fun :D  

[45910 rows x 9 columns]

In [32]:
df['duplica'] = df['user_id'] + df['review'] + df['item_id'] + df['funny'] + df['posted'] + df['last_edited'] + df['helpful'] + str(df['recommend'])
md.indentificar_duplicados(df,'duplica').sort_values(by=['user_id','item_id'])

user_id                                  user_url  \
13411     05041129     http://steamcommunity.com/id/05041129   
31985     05041129     http://steamcommunity.com/id/05041129   
13412     05041129     http://steamcommunity.com/id/05041129   
31986     05041129     http://steamcommunity.com/id/05041129   
13410     05041129     http://steamcommunity.com/id/05041129   
...            ...                                       ...   
43716  yolofaceguy  http://steamcommunity.com/id/yolofaceguy   
6752   yolofaceguy  http://steamcommunity.com/id/yolofaceguy   
43717  yolofaceguy  http://steamcommunity.com/id/yolofaceguy   
14835    zeroblade    http://steamcommunity.com/id/zeroblade   
34438    zeroblade    http://steamcommunity.com/id/zeroblade   

                                  funny                     posted  \
13411  1 person found this review funny    Posted January 5, 2015.   
31985  1 person found this review funny    Posted January 5, 2015.   
13412                                         Posted May 23, 2015.   
31986                                         Posted May 23, 2015.   
13410                                         Posted May 18, 2015.   
...                                 ...                        ...   
43716                                     Posted October 31, 2014.   
6752                                     Posted September 2, 2014.   
43717                                    Posted September 2, 2014.   
14835                                    Posted November 30, 2013.   
34438                                    Posted November 30, 2013.   

      last_edited item_id                                         helpful  \
13411              277430  1 of 1 people (100%) found this review helpful   
31985              277430  1 of 1 people (100%) found this review helpful   
13412                 440   1 of 2 people (50%) found this review helpful   
31986                 440   1 of 2 people (50%) found this review helpful   
13410                 730  4 of 4 people (100%) found this review helpful   
...           ...     ...                                             ...   
43716              250320                                  No ratings yet   
6752               261030                                  No ratings yet   
43717              261030                                  No ratings yet   
14835              219640   2 of 5 people (40%) found this review helpful   
34438              219640   2 of 5 people (40%) found this review helpful   

       recommend                                             review  \
13411       True  this is the best third person game ever that i...   
31985       True  this is the best third person game ever that i...   
13412      False  this will be the  number one game if it have m...   
31986      False  this will be the  number one game if it have m...   
13410       True  This game to me it is so good that it is bette...   
...          ...                                                ...   
43716       True  from the creaters of the walking dead, i prese...   
6752        True  this game is awesome,this game is ♥♥♥♥ed up an...   
43717       True  this game is awesome,this game is ♥♥♥♥ed up an...   
14835       True  SAVE THE KING!This is one of my favourite LAN ...   
34438       True  SAVE THE KING!This is one of my favourite LAN ...   

                                                 duplica  
13411  05041129this is the best third person game eve...  
31985  05041129this is the best third person game eve...  
13412  05041129this will be the  number one game if i...  
31986  05041129this will be the  number one game if i...  
13410  05041129This game to me it is so good that it ...  
...                                                  ...  
43716  yolofaceguyfrom the creaters of the walking de...  
6752   yolofaceguythis game is awesome,this game is ♥...  
43717  yolofaceguythis game is awesome,this game is ♥...  
14835  zerobladeSAVE THE KING!Thi

Ahora si, si son del mismo usuario y el review es exactamente igual, podemos eliminar los registros.

In [33]:
df.drop_duplicates(inplace=True)
md.indentificar_duplicados(df,'duplica').sort_values(by=['user_id','item_id'])

Empty DataFrame
Columns: [user_id, user_url, funny, posted, last_edited, item_id, helpful, recommend, review, duplica]
Index: []

In [34]:
df.shape


(58431, 10)

### Seleccion columnas

In [35]:
columnas = ['user_id','item_id','recommend','review','posted']

df = df[columnas]
df

user_id item_id  recommend  \
0      76561197970982479    1250       True   
1      76561197970982479   22200       True   
2      76561197970982479   43110       True   
3                js41637  251610       True   
4                js41637  227300       True   
...                  ...     ...        ...   
59300  76561198312638244      70       True   
59301  76561198312638244  362890       True   
59302        LydiaMorley  273110       True   
59303        LydiaMorley     730       True   
59304        LydiaMorley     440       True   

                                                  review  \
0      Simple yet with great replayability. In my opi...   
1                   It's unique and worth a playthrough.   
2      Great atmosphere. The gunplay can be a bit chu...   
3      I know what you think when you see this title ...   
4      For a simple (it's actually not all that simpl...   
...                                                  ...   
59300  a must have classic from steam definitely wort...   
59301  this game is a perfect remake of the original ...   
59302  had so much fun plaing this and collecting res...   
59303                                                 :D   
59304                                     so much fun :D   

                          posted  
0       Posted November 5, 2011.  
1          Posted July 15, 2011.  
2         Posted April 21, 2011.  
3          Posted June 24, 2014.  
4      Posted September 8, 2013.  
...                          ...  
59300            Posted July 10.  
59301             Posted July 8.  
59302             Posted July 3.  
59303            Posted July 20.  
59304             Posted July 2.  

[58431 rows x 5 columns]

### Tipo de datos

In [36]:
md.tipo_datos(df)

Columna Tipo_datos  %_nulos  Nulos
0    user_id      [str]     0.00      0
1    item_id      [str]     0.00      0
2  recommend     [bool]     0.00      0
3     review      [str]     0.00      0
4     posted      [str]     0.00      0

In [50]:
from importlib import reload
reload(md)
md.porcentaje_repr(dic['user_reviews'],'recommend')

Valor  Cantidad  Porcentaje
0   True     51692       88.47
1  False      6739       11.53

In [37]:
dic['user_reviews'] = df.copy()

## Users_Items

In [38]:
df = dic['users_items']
df

user_id  _count           steam_id  \
0        76561197970982479     277  76561197970982479   
1        76561197970982479     277  76561197970982479   
2        76561197970982479     277  76561197970982479   
3        76561197970982479     277  76561197970982479   
4        76561197970982479     277  76561197970982479   
...                    ...     ...                ...   
5153204  76561198329548331       7  76561198329548331   
5153205  76561198329548331       7  76561198329548331   
5153206  76561198329548331       7  76561198329548331   
5153207  76561198329548331       7  76561198329548331   
5153208  76561198329548331       7  76561198329548331   

                                                  user_url item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10   
1        http://steamcommunity.com/profiles/76561197970...      20   
2        http://steamcommunity.com/profiles/76561197970...      30   
3        http://steamcommunity.com/profiles/76561197970...      40   
4        http://steamcommunity.com/profiles/76561197970...      50   
...                                                    ...     ...   
5153204  http://steamcommunity.com/profiles/76561198329...  346330   
5153205  http://steamcommunity.com/profiles/76561198329...  373330   
5153206  http://steamcommunity.com/profiles/76561198329...  388490   
5153207  http://steamcommunity.com/profiles/76561198329...  521570   
5153208  http://steamcommunity.com/profiles/76561198329...  519140   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike                 6                0  
1                Team Fortress Classic                 0                0  
2                        Day of Defeat                 7                0  
3                   Deathmatch Classic                 0                0  
4            Half-Life: Opposing Force                 0                0  
...                                ...               ...              ...  
5153204                   BrainBread 2                 0                0  
5153205                    All Is Dust                 0                0  
5153206  One Way To Die: Steam Edition                 3                3  
5153207          You Have 10 Seconds 2                 4                4  
5153208                     Minds Eyes                 3                3  

[5153209 rows x 8 columns]

### Registros nulos

In [39]:
# eliminemos los registros nulos
print(df.shape)
df = df.dropna(how='all').reset_index(drop=True)
print(df.shape) # no hay registros nulos completamente

(5153209, 8)
(5153209, 8)


### Registros duplicados

In [40]:
md.indentificar_duplicados(df,'user_id') 
# como era de esperarse, tenemos el mismo problema que anteriormente

user_id  _count           steam_id  \
0        76561197970982479     277  76561197970982479   
1        76561197970982479     277  76561197970982479   
2        76561197970982479     277  76561197970982479   
3        76561197970982479     277  76561197970982479   
4        76561197970982479     277  76561197970982479   
...                    ...     ...                ...   
5153204  76561198329548331       7  76561198329548331   
5153205  76561198329548331       7  76561198329548331   
5153206  76561198329548331       7  76561198329548331   
5153207  76561198329548331       7  76561198329548331   
5153208  76561198329548331       7  76561198329548331   

                                                  user_url item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10   
1        http://steamcommunity.com/profiles/76561197970...      20   
2        http://steamcommunity.com/profiles/76561197970...      30   
3        http://steamcommunity.com/profiles/76561197970...      40   
4        http://steamcommunity.com/profiles/76561197970...      50   
...                                                    ...     ...   
5153204  http://steamcommunity.com/profiles/76561198329...  346330   
5153205  http://steamcommunity.com/profiles/76561198329...  373330   
5153206  http://steamcommunity.com/profiles/76561198329...  388490   
5153207  http://steamcommunity.com/profiles/76561198329...  521570   
5153208  http://steamcommunity.com/profiles/76561198329...  519140   

                             item_name  playtime_forever  playtime_2weeks  
0                       Counter-Strike                 6                0  
1                Team Fortress Classic                 0                0  
2                        Day of Defeat                 7                0  
3                   Deathmatch Classic                 0                0  
4            Half-Life: Opposing Force                 0                0  
...                                ...               ...              ...  
5153204                   BrainBread 2                 0                0  
5153205                    All Is Dust                 0                0  
5153206  One Way To Die: Steam Edition                 3                3  
5153207          You Have 10 Seconds 2                 4                4  
5153208                     Minds Eyes                 3                3  

[5149907 rows x 8 columns]

In [41]:
df['duplica'] = df.astype(str).sum(axis=1)
df

user_id  _count           steam_id  \
0        76561197970982479     277  76561197970982479   
1        76561197970982479     277  76561197970982479   
2        76561197970982479     277  76561197970982479   
3        76561197970982479     277  76561197970982479   
4        76561197970982479     277  76561197970982479   
...                    ...     ...                ...   
5153204  76561198329548331       7  76561198329548331   
5153205  76561198329548331       7  76561198329548331   
5153206  76561198329548331       7  76561198329548331   
5153207  76561198329548331       7  76561198329548331   
5153208  76561198329548331       7  76561198329548331   

                                                  user_url item_id  \
0        http://steamcommunity.com/profiles/76561197970...      10   
1        http://steamcommunity.com/profiles/76561197970...      20   
2        http://steamcommunity.com/profiles/76561197970...      30   
3        http://steamcommunity.com/profiles/76561197970...      40   
4        http://steamcommunity.com/profiles/76561197970...      50   
...                                                    ...     ...   
5153204  http://steamcommunity.com/profiles/76561198329...  346330   
5153205  http://steamcommunity.com/profiles/76561198329...  373330   
5153206  http://steamcommunity.com/profiles/76561198329...  388490   
5153207  http://steamcommunity.com/profiles/76561198329...  521570   
5153208  http://steamcommunity.com/profiles/76561198329...  519140   

                             item_name  playtime_forever  playtime_2weeks  \
0                       Counter-Strike                 6                0   
1                Team Fortress Classic                 0                0   
2                        Day of Defeat                 7                0   
3                   Deathmatch Classic                 0                0   
4            Half-Life: Opposing Force                 0                0   
...                                ...               ...              ...   
5153204                   BrainBread 2                 0                0   
5153205                    All Is Dust                 0                0   
5153206  One Way To Die: Steam Edition                 3                3   
5153207          You Have 10 Seconds 2                 4                4   
5153208                     Minds Eyes                 3                3   

                                                   duplica  
0        7656119797098247927776561197970982479http://st...  
1        7656119797098247927776561197970982479http://st...  
2        7656119797098247927776561197970982479http://st...  
3        7656119797098247927776561197970982479http://st...  
4        7656119797098247927776561197970982479http://st...  
...                                                    ...  
5153204  76561198329548331776561198329548331http://stea...  
5153205  76561198329548331776561198329548331http://stea...  
5153206  76561198329548331776561198329548331http://stea...  
5153207  76561198329548331776561198329548331http://stea...  
5153208  76561198329548331776561198329548331http://stea...  

[5153209 rows x 9 columns]

In [42]:
md.indentificar_duplicados(df,'duplica')

user_id  _count           steam_id  \
4344     76561198156664158      59  76561198156664158   
4345     76561198156664158      59  76561198156664158   
4346     76561198156664158      59  76561198156664158   
4347     76561198156664158      59  76561198156664158   
4348     76561198156664158      59  76561198156664158   
...                    ...     ...                ...   
4898223  76561198080057659      39  76561198080057659   
4898224  76561198080057659      39  76561198080057659   
4898225  76561198080057659      39  76561198080057659   
4898226  76561198080057659      39  76561198080057659   
4898227  76561198080057659      39  76561198080057659   

                                                  user_url item_id  \
4344     http://steamcommunity.com/profiles/76561198156...    4000   
4345     http://steamcommunity.com/profiles/76561198156...   33910   
4346     http://steamcommunity.com/profiles/76561198156...   33930   
4347     http://steamcommunity.com/profiles/76561198156...  219540   
4348     http://steamcommunity.com/profiles/76561198156...     400   
...                                                    ...     ...   
4898223  http://steamcommunity.com/profiles/76561198080...  213670   
4898224  http://steamcommunity.com/profiles/76561198080...  221910   
4898225  http://steamcommunity.com/profiles/76561198080...  261030   
4898226  http://steamcommunity.com/profiles/76561198080...  273110   
4898227  http://steamcommunity.com/profiles/76561198080...     730   

                                           item_name  playtime_forever  \
4344                                     Garry's Mod               269   
4345                                          Arma 2               162   
4346                     Arma 2: Operation Arrowhead               223   
4347     Arma 2: Operation Arrowhead Beta (Obsolete)                 0   
4348                                          Portal               196   
...                                              ...               ...   
4898223             South Park™: The Stick of Truth™               725   
4898224                          The Stanley Parable                53   
4898225                 The Walking Dead: Season Two               253   
4898226                Counter-Strike Nexon: Zombies                 0   
4898227             Counter-Strike: Global Offensive                 0   

         playtime_2weeks                                            duplica  
4344                   0  765611981566641585976561198156664158http://ste...  
4345                   0  765611981566641585976561198156664158http://ste...  
4346                   0  765611981566641585976561198156664158http://ste...  
4347                   0  765611981566641585976561198156664158http://ste...  
4348                   0  765611981566641585976561198156664158http://ste...  
...                  ...                                                ...  
4898223                0  765611980800576593976561198080057659http://ste...  
4898224                0  765611980800576593976561198080057659http://ste...  
4898225                0  765611980800576593976561198080057659http://ste...  
4898226                0  765611980800576593976561198080057659http://ste...  
4898227                0  765611980800576593976561198080057659http://ste...  

[117731 rows x 9 columns]

In [43]:
df.drop_duplicates(inplace=True)
md.indentificar_duplicados(df,'duplica')

Empty DataFrame
Columns: [user_id, _count, steam_id, user_url, item_id, item_name, playtime_forever, playtime_2weeks, duplica]
Index: []

### Seleccion Columnas

In [44]:
columnas = ['user_id','steam_id','item_id','item_name','playtime_forever']

df = df[columnas]
df

user_id           steam_id item_id  \
0        76561197970982479  76561197970982479      10   
1        76561197970982479  76561197970982479      20   
2        76561197970982479  76561197970982479      30   
3        76561197970982479  76561197970982479      40   
4        76561197970982479  76561197970982479      50   
...                    ...                ...     ...   
5153204  76561198329548331  76561198329548331  346330   
5153205  76561198329548331  76561198329548331  373330   
5153206  76561198329548331  76561198329548331  388490   
5153207  76561198329548331  76561198329548331  521570   
5153208  76561198329548331  76561198329548331  519140   

                             item_name  playtime_forever  
0                       Counter-Strike                 6  
1                Team Fortress Classic                 0  
2                        Day of Defeat                 7  
3                   Deathmatch Classic                 0  
4            Half-Life: Opposing Force                 0  
...                                ...               ...  
5153204                   BrainBread 2                 0  
5153205                    All Is Dust                 0  
5153206  One Way To Die: Steam Edition                 3  
5153207          You Have 10 Seconds 2                 4  
5153208                     Minds Eyes                 3  

[5094105 rows x 5 columns]

### Tipo de datos

In [45]:
md.tipo_datos(df)

Columna Tipo_datos  %_nulos  Nulos
0           user_id      [str]     0.00      0
1          steam_id      [str]     0.00      0
2           item_id      [str]     0.00      0
3         item_name      [str]     0.00      0
4  playtime_forever      [int]     0.00      0

#### Estandarizacion de tipeo

In [46]:
text_cols = ['item_name']

df['item_name'] = df['item_name'].apply(md.convertir_proper)


C:\Users\pedro\AppData\Local\Temp\ipykernel_5752\568049889.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[text_cols] = df[text_cols].apply(md.convertir_proper)


In [47]:
dic['users_items'] = df.copy()

# Guardado (Load)

In [48]:
for k,_ in dic.items():
    dic[k].to_parquet(nombres[k][1],index=False)